### PR TITLE
Put missing _none and _some filters into GraphQL API.

### LINKED ISSUE #14228: _none and _some seem to be missing in GraphQL
### Describe the Bug

It seems that _none and _some filters are not available in GraphQL. 

### To Reproduce

Try a query like the one described [here](https://docs.directus.io/reference/filter-rules/#some-vs-none-in-one-to-many)

### Errors Shown

Field \"_none\" is not defined by type \"__REDACTED__\". Did you mean \"_and\" or \"_or\"?"

### What version of Directus are you using?

9.11.1

### What version of Node.js are you using?

official docker image

### What database are you using?

mariadb

### What browser are you using?

Chrome

### How are you deploying Directus?

Docker

### PR DESCRIPTION
**Pull request by freekrai**
## Description

The GraphQL API is missing the `_none` and `_some` filters, which causes errors if you attempt to use them. 

This PR puts them in.

Fixes #14228

## Type of Change

- [X] Bugfix
- [ ] Improvement
- [ ] New Feature
- [ ] Refactor / codestyle updates
- [ ] Other, please describe:

## Requirements Checklist

- [ ] New / updated tests are included
- [X] All tests are passing locally
- [X] Performed a self-review of the submitted code

If adding a new feature:

- [ ] Documentation was added/updated

### CODE REVIEW

File: api/src/services/graphql/index.ts [hunk 1]
```diff
@@ -851,6 +851,8 @@ export class GraphQLService {
 				ReadableCollectionFilterTypes[collection.collection].addFields({
 					_and: [ReadableCollectionFilterTypes[collection.collection]],
 					_or: [ReadableCollectionFilterTypes[collection.collection]],
+					_none: [ReadableCollectionFilterTypes[collection.collection]],
+					_some: [ReadableCollectionFilterTypes[collection.collection]],
```
Comment (rijkvanzanten) [inline api/src/services/graphql/index.ts:855]: `none` and `some` are only available as the first level nesting of a O2M field, like:

```
{
  _and: [
    {
      title: {
        _eq: 'Example'
      }
    },
    {
      favorites: {
        _none: { <<<<<<<<<<<<<<<
          rating: {
            _lte: 3
          }
        }
      }
    }
  ]
}

**Review by rijkvanzanten — CHANGES_REQUESTED** [review]

File: api/src/services/graphql/index.ts (hunk 1 above)
Comment (freekrai) [reply api/src/services/graphql/index.ts:855]: Ok, will come back to this one and make sure we're only using it in the right place after the I push the other PR 👍🏻

**Comment by freekrai** [comment]
@rijkvanzanten can I get a second look at this PR please? where I've moved the `_some` and `_none` only works during relationships now

**Review by licitdev — CHANGES_REQUESTED** [review]
Tested the PR and some findings as follows.

The `_none` and `_some` filter operators work on the top level filter as expected.

```gql
{
  restaurant(filter: {dishes: {_none: {available: {_eq: false}}}}) {
    name
    dishes {
      available
    }
  }
}
```

However, the operators should not be usable on nested fields.

image

Additionally, they should only be usable on the first level after a relational field. The following query should fail the GraphQL validation.

```gql
{
  restaurant(
    filter: {dishes: {_none: {_none: {_none: {available: {_eq: false}}}}}}
  ) {
    name
  }
}
```

Finally, the operators should only be added for relational fields that returns an array of results (eg: o2m).

image

**Comment by freekrai** [comment]
Ok, will revisit, probably need to add a check for relationship level

**Comment by freekrai** [comment]
Ok, will revisit, probably need to add a check for relationship level so it's applied only at the first level and not further down.

**Comment by freekrai** [comment]
just enforced that `_some` and `_none` only work on relationships.

Testing 

```
{
  restaurant(
    filter: {dishes: {_none: {_none: {_none: {available: {_eq: false}}}}}}
  ) {
    name
  }
}
```

already fails validation for me which we want obviously.

Whereas

```
query {
    restaurant(
        filter: {
            dishes: {
                _some: {
                    available: {
                        _eq: false
                    }
                }
            }
        }
    ) {
        name
        dishes {
            name
            available
        }
  }
}
```

passes as expected.

**Comment by licitdev** [comment]
> already fails validation for me

The first query is passing for me though, validation wise. The error thrown is a GraphQL validation error. 🤔

image

The second query passes as well.

**Comment by freekrai** [comment]
Yeah it should fail with validation so will see why schema lets it pass.

**Comment by freekrai** [comment]
So currently, trying to add `_some` or `_none` outside the proper collection will fail validation.

Screen Shot 2022-09-19 at 10 30 11 AM

Working on now making it only allow it to pass on the first round, rather than letting you enter `_none: { _none: {...}}`` which fails to run due to being invalid but the schema still lets you enter it.

**Comment by freekrai** [comment]
I've added a check if `_some` / `_none` appear beyond top level, but otherwise, there's not much else I can see we can do here, it won't run if you use `_some` / `_none` beyond top level, but the schema will display it as an option inside `dishes` for example due to being part of that filter.

@licitdev unless you have ideas?

File: api/src/utils/validate-query.ts [hunk 2]
```diff
@@ -5,6 +5,7 @@ import { Query } from '@directus/shared/types';
 import { stringify } from 'wellknown';
 import { calculateFieldDepth } from './calculate-field-depth';
 import env from '../env';
+import e from 'express';
```
Comment (licitdev) [inline api/src/utils/validate-query.ts:8]:
```suggestion
```
Fix lint

File: api/src/utils/validate-query.ts [hunk 3]
```diff
@@ -101,6 +100,61 @@ function validateFilter(filter: Query['filter']) {
 	}
 }
 
+function validateNestedFilter(filter: Query['filter']) {
```
Comment (licitdev) [inline api/src/utils/validate-query.ts:103]: I think that `validateFilter` should be refactored to accept a boolean such as `isNested`, as `validateNestedFilter` essentially the same with the removal of `_none` and `_some`.

File: api/src/utils/validate-query.ts [hunk 4]
```diff
@@ -42,12 +42,14 @@ export function validateQuery(query: Query): Query {
 	return query;
 }
 
-function validateFilter(filter: Query['filter']) {
+function validateFilter(filter: Query['filter'], isNested = false) {
 	if (!filter) throw new InvalidQueryException('Invalid filter object');
-
 	for (const [key, nested] of Object.entries(filter)) {
+		if (key === '_some' || key === '_none') {
+			if (isNested) throw new InvalidQueryException('Invalid filter object');
```
Comment (licitdev) [inline api/src/utils/validate-query.ts:49]: I think this error can be more specific along the lines of invalid usage of `_some` or `_none`.

**Review by licitdev — CHANGES_REQUESTED** [review]
There are still some inconsistencies.

image

Line 2:
`_some` working as expected as `ingredients` is an `o2m` field

Line 5:
`_some` should not be usable on the top level.
This can't be done on `food`, but it is on `ingredients`.
ie: `_some` is in `ingredients_type` but not in `food_type`

Line 8 & Line 11:
Multiple usages of `_some` and `_none` is caught in `validateFilter` but not in GraphQL validation. 
@freekrai Can this behaviour be fixed?
@rijkvanzanten is it acceptable for GraphQL validation to pass but only fail when running the request?

**Comment by rijkvanzanten** [comment]
> @rijkvanzanten is it acceptable for GraphQL validation to pass but only fail when running the request?

If there's no other option, then yes. Ideally it matches up though

**Comment by freekrai** [comment]
> > @rijkvanzanten is it acceptable for GraphQL validation to pass but only fail when running the request?
> 
> If there's no other option, then yes. Ideally it matches up though

will look at making sure everything matches up elsewhere but otherwise, yeah, GraphQL will validate it during introspection, it just won’t let it pass when you run the query.

**Comment by licitdev** [comment]
@freekrai Marking this PR as draft for now as there are still inconsistencies to be sorted out. Will review it again after it's marked as ready. Thanks! 👍

**Comment by freekrai** [comment]
> There are still some inconsistencies.
> 
> image
> 
> Line 2: `_some` working as expected as `ingredients` is an `o2m` field
> 
> Line 5: `_some` should not be usable on the top level. This can't be done on `food`, but it is on `ingredients`. ie: `_some` is in `ingredients_type` but not in `food_type`
> 
> Line 8 & Line 11: Multiple usages of `_some` and `_none` is caught in `validateFilter` but not in GraphQL validation. @freekrai Can this behaviour be fixed? @rijkvanzanten is it acceptable for GraphQL validation to pass but only fail when running the request?

I’ll look at line 5 but 8 and 11, no, there’s not much we can do as that is where the most time has been lost on this PR.

**Comment by freekrai** [comment]
> > There are still some inconsistencies.
> > image
> > Line 2: `_some` working as expected as `ingredients` is an `o2m` field
> > Line 5: `_some` should not be usable on the top level. This can't be done on `food`, but it is on `ingredients`. ie: `_some` is in `ingredients_type` but not in `food_type`
> > Line 8 & Line 11: Multiple usages of `_some` and `_none` is caught in `validateFilter` but not in GraphQL validation. @freekrai Can this behaviour be fixed? @rijkvanzanten is it acceptable for GraphQL validation to pass but only fail when running the request?
> 
> I’ll look at line 5 but 8 and 11, no, there’s not much we can do as that is where the most time has been lost on this PR.

For 5, when I try to run it:

```
query {
  dishes(
        filter: {
            _some: {
                available: {
                    _eq: false
                }
            }
        }
    ) {
        name
        available
  }
}
```

I get an error:

```
{
  "errors": [
    {
      "message": "\"dishes._some\" is not a relational field",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "dishes"
      ]
    }
  ],
  "data": null
}
```

since we only even add _some / _none to O2M relationships, so that comes back to the introspection where it allows _some / _none even if it won't run 🤔

**Comment by rijkvanzanten** [comment]
What's left to be done on this?

**Comment by freekrai** [comment]
> What's left to be done on this?

Not sure. Ideally we’d get _some / _none only appearing during introspection at certain points but haven’t been able to do that. So right now it’ll throw an error if you try it when not supposed to and only allow _some / _none when in called on an O2M field and not nested.

**Comment by rijkvanzanten** [comment]
Is that ideally something we're aiming to implement in this PR, or is this ready for review as is?

**Comment by freekrai** [comment]
It’s ready for review again as not sure I can implement it in this PR, as it’s where I’ve lost a lot of time on this PR trying to implement it. It’ll give an error if you try to use _some / _none nested or where not allowed, it’s just introspection will say it is allowed before it runs, so probably some more deep dives in validating that mean a part 2 at some point.

**Comment by rijkvanzanten** [comment]
Lets get some tests in place for these changes and I think we're good to go 🙂

**Comment by licitdev** [comment]
@rijkvanzanten I'll be re-reviewing this PR _soon_, after base GraphQL tests are done. 
I will be adding tests for this PR too. :)

**Comment by rijkvanzanten** [comment]
Closing this to revisit soon

**Pull request closed without merging**